File: Readers/ReaderFactory.cpp
// Written by Navdeep Jaitly for the Department of Energy (PNNL, Richland, WA)
// Copyright 2006, Battelle Memorial Institute
// E-mail: navdeep.jaitly@pnl.gov
// Website: http://ncrr.pnl.gov/software
// -------------------------------------------------------------------------------
// 
// Licensed under the Apache License, Version 2.0; you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at 
// http://www.apache.org/licenses/LICENSE-2.0

#include "ReaderFactory.h"
#include "BrukerRawData.h"
#include "BrukerAsciiRawData.h"
#include "AsciiRawData.h"
#include "SunExtrelRawData.h"
#include "FinniganRawData.h" 
#include "AgilentRawData.h" 
#include "Icr2lsRawData.h" 
#include "MicromassRawData.h"
#include "MZXmlRawData.h" 
#include "IMSRawData.h" 
#include "UIMFRawData.h" 
#include "YafmsRawData.h"

namespace Engine 
{
	namespace Readers
	{
		RawData* ReaderFactory::GetRawData(FileType file_type, char *file_name)
		{
			char *header_n = "header" ; 
			switch(file_type)
			{
				case BRUKER:
					BrukerRawData *bruker_raw_data ; 
					bruker_raw_data = new BrukerRawData() ; 
					bruker_raw_data->Load(file_name) ; 
					return bruker_raw_data ; 
					break ; 
				case BRUKER_ASCII:
					BrukerAsciiRawData *bruker_baf_raw_data ; 
					bruker_baf_raw_data = new BrukerAsciiRawData() ; 
					bruker_baf_raw_data->Load(file_name) ; 
					return bruker_baf_raw_data ; 
					break ; 
				case ASCII:
					AsciiRawData *ascii_raw_data ; 
					ascii_raw_data = new AsciiRawData() ; 
					ascii_raw_data->Load(file_name) ; 
					return ascii_raw_data ; 
					break ; 
				case AGILENT_TOF:
#ifdef ANALYST_INSTALLED
					AgilentRawData *agilent_raw_data ; 
					agilent_raw_data = new AgilentRawData() ; 
					agilent_raw_data->Load(file_name) ; 
					return agilent_raw_data ; 
#endif 
					break ; 
				case SUNEXTREL:
					SunExtrelRawData *sun_extrel_raw_data  ;
					sun_extrel_raw_data = new SunExtrelRawData() ; 
					sun_extrel_raw_data->Load(file_name) ; 
					return sun_extrel_raw_data ; 
					break ; 
				case ICR2LSRAWDATA:
					Icr2lsRawData *icr2ls_raw_data  ;
					icr2ls_raw_data = new Icr2lsRawData() ; 
					icr2ls_raw_data->Load(file_name) ; 
					return icr2ls_raw_data ; 
					break ; 
				case FINNIGAN:
#ifdef XCALIBUR_INSTALLED
					FinniganRawData *finnigan_raw_data ; 
					finnigan_raw_data = new FinniganRawData() ;
					finnigan_raw_data->Load(file_name) ; 
					return finnigan_raw_data ; 
#endif 
					break ; 
				case MICROMASSRAWDATA:
					MicromassRawData *micro_raw_data ; 
					micro_raw_data = new MicromassRawData() ;
					micro_raw_data->Load(file_name) ; 
					return micro_raw_data ; 
					break ; 
				case MZXMLRAWDATA:
					MZXmlRawData *mzxml_raw_data ; 
					mzxml_raw_data = new MZXmlRawData() ; 
					mzxml_raw_data->Load(file_name) ; 
					return mzxml_raw_data ; 
					break ; 
				case PNNL_IMS:
					IMSRawData *ims_raw_data ; 
					ims_raw_data = new IMSRawData() ; 
					ims_raw_data->Load(file_name) ; 
					return ims_raw_data ; 
					break ; 
				case PNNL_UIMF:
					UIMFRawData *uimf_raw_data ; 
					uimf_raw_data = new UIMFRawData() ; 
					uimf_raw_data->Load(file_name) ; 
					return uimf_raw_data ; 
					break ; 
				case YAFMS:
					YafmsRawData *yafms_raw_data ; 
					yafms_raw_data = new YafmsRawData() ; 
					yafms_raw_data->Load(file_name) ; 
					return yafms_raw_data ; 
					break ; 
				default:
					break ; 
			}
			return NULL ; 
		}

		void ReaderFactory::GetRawData(RawData **raw_data, FileType file_type)
		{
			char *header_n = "header" ; 
			*raw_data = NULL ; 
			switch(file_type)
			{
				case BRUKER:
					*raw_data = new BrukerRawData() ; 
					return ; 
					break ; 
				case AGILENT_TOF:
#ifdef ANALYST_INSTALLED
					*raw_data = new AgilentRawData() ; 
					return ; 
#endif 
					break ; 
				case FINNIGAN:
#ifdef XCALIBUR_INSTALLED
					*raw_data = new FinniganRawData() ; 
					return ; 
#endif 
					break ; 
				case SUNEXTREL:
					*raw_data = new SunExtrelRawData() ;
					return ; 
					break ; 
				case ICR2LSRAWDATA:
					*raw_data = new Icr2lsRawData() ; 
					return ; 
					break ; 
				case MICROMASSRAWDATA:
					*raw_data = new MicromassRawData() ; 
					return ; 
					break ; 
				case MZXMLRAWDATA:
					*raw_data = new MZXmlRawData() ; 
					return ; 
					break ; 
				case PNNL_IMS:
					*raw_data = new IMSRawData() ; 
					return ;
					break ; 
				case PNNL_UIMF:
					*raw_data = new UIMFRawData() ; 
					return ;
					break ; 
				case YAFMS:
					*raw_data = new YafmsRawData() ; 
					return ;
					break ; 
				case BRUKER_ASCII:
					*raw_data = new BrukerAsciiRawData() ; 
					return ; 
					break ; 
				case ASCII:
					*raw_data = new AsciiRawData() ; 
					return ; 
					break ; 
				default:
					break ; 
			}
			return  ; 
		}

		RawData* ReaderFactory::GetRawData(FileType file_type)
		{
			char *header_n = "acqu" ; 
			switch(file_type)
			{
				case BRUKER:
					BrukerRawData *bruker_raw_data ; 
					bruker_raw_data = new BrukerRawData() ; 
					return bruker_raw_data ; 
					break ; 
				case AGILENT_TOF:
#ifdef ANALYST_INSTALLED
					AgilentRawData *agilent_raw_data ; 
					agilent_raw_data= new AgilentRawData() ; 
					return agilent_raw_data ; 
#endif
					break ; 
				case FINNIGAN:
#ifdef XCALIBUR_INSTALLED
					FinniganRawData *finnigan_raw_data ; 
					finnigan_raw_data = new FinniganRawData() ;
					return finnigan_raw_data ; 
#endif 
					break ; 
				case SUNEXTREL:
					SunExtrelRawData *sun_extrel_raw_data  ;
					sun_extrel_raw_data = new SunExtrelRawData() ; 
					return sun_extrel_raw_data ; 
					break ; 
				case ICR2LSRAWDATA:
					Icr2lsRawData *icr2ls_raw_data  ;
					icr2ls_raw_data = new Icr2lsRawData() ; 
					return icr2ls_raw_data ; 
					break ; 
				case MICROMASSRAWDATA:
					MicromassRawData *micro_raw_data ; 
					micro_raw_data = new MicromassRawData() ; 
					return micro_raw_data ; 
					break ; 
				case MZXMLRAWDATA:
					MZXmlRawData *mzxml_raw_data ; 
					mzxml_raw_data = new MZXmlRawData() ; 
					return mzxml_raw_data ; 
					break ; 
				case PNNL_IMS:
					IMSRawData *ims_raw_data ; 
					ims_raw_data = new IMSRawData() ; 
					return ims_raw_data ; 
					break ; 
				case PNNL_UIMF:
					UIMFRawData *uimf_raw_data ; 
					uimf_raw_data = new UIMFRawData() ; 
					return uimf_raw_data ; 
					break ; 
				case YAFMS:
					YafmsRawData *yafms_raw_data ; 
					yafms_raw_data = new YafmsRawData() ; 
					return yafms_raw_data ; 
					break ; 
				case BRUKER_ASCII:
					BrukerAsciiRawData *bruker_baf_raw_data ; 
					bruker_baf_raw_data = new BrukerAsciiRawData() ; 
					return bruker_baf_raw_data ; 
					break ; 
				case ASCII:
					AsciiRawData *ascii_raw_data ; 
					ascii_raw_data = new AsciiRawData() ; 
					return ascii_raw_data ; 
					break ; 
				default:
					break ; 
			}
			return NULL ; 
		}
	}
}

File: C#_Version/DeconEngine/Readers/ReaderFactory.cs
using System;
using DeconToolsV2.Readers;

namespace Engine.Readers
{
    internal class ReaderFactory
    {
#if Enable_Obsolete
        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public static RawData GetMSDataReader(FileType fileType)
        {
            return GetRawData(fileType);
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public static RawData GetMSDataReader(FileType fileType, string fileName)
        {
            return GetRawData(fileType, fileName);
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public static void GetMSDataReader(out RawData rawData, FileType fileType)
        {
            GetRawData(out rawData, fileType);
        }
#endif

        public static RawData GetRawData(FileType fileType, string fileName)
        {
            RawData rawData = null;
            switch (fileType)
            {
#if Enable_Obsolete
                case FileType.BRUKER:
                    rawData = new BrukerRawData();
                    rawData.Load(fileName);
                    break;
                case FileType.BRUKER_ASCII:
                    rawData = new BrukerAsciiRawData();
                    rawData.Load(fileName);
                    break;
                case FileType.ASCII:
                    rawData = new AsciiRawData();
                    rawData.Load(fileName);
                    break;
                case FileType.AGILENT_TOF:
#if ANALYST_INSTALLED
                    rawData = new AgilentRawData();
                    rawData.Load(fileName);
#endif
                    break;
                case FileType.SUNEXTREL:
                    rawData = new SunExtrelRawData();
                    rawData.Load(fileName);
                    break;
                case FileType.ICR2LSRAWDATA:
                    rawData = new Icr2lsRawData();
                    rawData.Load(fileName);
                    break;
#endif
                case FileType.FINNIGAN:
#if XCALIBUR_INSTALLED
                    rawData = new FinniganRawData();
                    rawData.Load(fileName);
#endif
                    break;
#if Enable_Obsolete
                case FileType.MICROMASSRAWDATA:
#if MASSLYNX_4_INSTALLED
                    rawData = new MicromassRawData();
                    rawData.Load(fileName);
#endif
                    break;
#endif
                case FileType.MZXMLRAWDATA:
                    rawData = new MZXmlRawData();
                    rawData.Load(fileName);
                    break;
#if Enable_Obsolete
                case FileType.PNNL_IMS:
                    rawData = new IMSRawData();
                    rawData.Load(fileName);
                    break;
#endif
                //case FileType.PNNL_UIMF:
                //    rawData = new UIMFRawData();
                //    rawData.Load(file_name);
                //    break;
                /*case FileType.YAFMS:
                    rawData = new YafmsRawData();
                    rawData.Load(file_name);
                    break; */
                default:
                    break;
            }
            return rawData;
        }

#if Enable_Obsolete
        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public static void GetRawData(out RawData rawData, FileType fileType)
        {
            rawData = null;
            switch (fileType)
            {
                case FileType.BRUKER:
                    rawData = new BrukerRawData();
                    break;
                case FileType.AGILENT_TOF:
#if ANALYST_INSTALLED
                    rawData = new AgilentRawData();
#endif
                    break;
                case FileType.FINNIGAN:
#if XCALIBUR_INSTALLED
                    rawData = new FinniganRawData();
#endif
                    break;
                case FileType.SUNEXTREL:
                    rawData = new SunExtrelRawData();
                    break;
                case FileType.ICR2LSRAWDATA:
                    rawData = new Icr2lsRawData();
                    break;
                case FileType.MICROMASSRAWDATA:
#if MASSLYNX_4_INSTALLED
                    rawData = new MicromassRawData();
#endif
                    break;
                case FileType.MZXMLRAWDATA:
                    rawData = new MZXmlRawData();
                    break;
                case FileType.PNNL_IMS:
                    rawData = new IMSRawData();
                    break;
                /*case FileType.PNNL_UIMF:
                    rawData = new UIMFRawData();
                    break; */
                /*case FileType.YAFMS:
                    rawData = new YafmsRawData();
                    break; */
                case FileType.BRUKER_ASCII:
                    rawData = new BrukerAsciiRawData();
                    break;
                case FileType.ASCII:
                    rawData = new AsciiRawData();
                    break;
                default:
                    break;
            }
        }

        [Obsolete("Only used by DeconTools for ICR2LSRun and IMFRun; BrukerV2 exists, but has no use path", false)]
        public static RawData GetRawData(FileType fileType)
        {
            RawData rawData = null;
            switch (fileType)
            {
                case FileType.BRUKER:
                    rawData = new BrukerRawData();
                    break;
                case FileType.AGILENT_TOF:
#if ANALYST_INSTALLED
                    rawData = new AgilentRawData();
#endif
                    break;
                case FileType.FINNIGAN:
#if XCALIBUR_INSTALLED
                    rawData = new FinniganRawData();
#endif
                    break;
                case FileType.SUNEXTREL:
                    rawData = new SunExtrelRawData();
                    break;
                case FileType.ICR2LSRAWDATA:
                    rawData = new Icr2lsRawData();
                    break;
                case FileType.MICROMASSRAWDATA:
#if MASSLYNX_4_INSTALLED
                    rawData = new MicromassRawData();
#endif
                    break;
                case FileType.MZXMLRAWDATA:
                    rawData = new MZXmlRawData();
                    break;
                case FileType.PNNL_IMS:
                    rawData = new IMSRawData();
                    break;
                /*case FileType.PNNL_UIMF:
                    rawData = new UIMFRawData();
                    break; */
                /*case FileType.YAFMS:
                    rawData = new YafmsRawData();
                    break; */
                case FileType.BRUKER_ASCII:
                    rawData = new BrukerAsciiRawData();
                    break;
                case FileType.ASCII:
                    rawData = new AsciiRawData();
                    break;
                default:
                    break;
            }
            return rawData;
        }
#endif
    }
}
